From 3129fafb1af8e8ca3a4fa0493dc31a2513a2c970 Mon Sep 17 00:00:00 2001
From: Dmitry Mikushin <dmitry@kernelgen.org>
Date: Wed, 2 Apr 2025 15:04:53 +0200
Subject: [PATCH] According to CMAKE_<LANG>_FLAGS design, it should provide
 flags that are common for all configurations. And then
 CMAKE_<LANG>_FLAGS_DEBUG et al should provide _specific_ _extra_ flags for
 each configuration. The _RTL_FLAGS does not follow this design by appending
 Windows DLL flags to each configuration directly. As a result, the users who
 legitimately overwrite CMAKE_<LANG>_FLAGS_DEBUG (because they expect the
 common part of flags to be preserved in CMAKE_<LANG>_FLAGS) will loose the
 Windows DLL flags and break the build. This patch corrects the _RTL_FLAGS
 behavior.

---
 Modules/Platform/Windows-Clang.cmake | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/Modules/Platform/Windows-Clang.cmake b/Modules/Platform/Windows-Clang.cmake
index 34ed2bcf91..ee468ba871 100644
--- a/Modules/Platform/Windows-Clang.cmake
+++ b/Modules/Platform/Windows-Clang.cmake
@@ -102,7 +102,7 @@ macro(__windows_compiler_clang_gnu lang)
       set(_RTL_FLAGS "")
       set(_RTL_FLAGS_DEBUG "")
     else()
-      set(_RTL_FLAGS_DEBUG " -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd")
+      set(_RTL_FLAGS_DEBUG " -D_DEBUG")
       set(_RTL_FLAGS " -D_DLL -D_MT -Xclang --dependent-lib=msvcrt")
     endif()
 
@@ -112,10 +112,11 @@ macro(__windows_compiler_clang_gnu lang)
       set(_DBG_FLAGS " -g -Xclang -gcodeview")
     endif()
 
+    string(APPEND CMAKE_${lang}_FLAGS " ${_RTL_FLAGS}")
     string(APPEND CMAKE_${lang}_FLAGS_DEBUG_INIT " -O0${_DBG_FLAGS}${_RTL_FLAGS_DEBUG}")
-    string(APPEND CMAKE_${lang}_FLAGS_MINSIZEREL_INIT " -Os -DNDEBUG${_RTL_FLAGS}")
-    string(APPEND CMAKE_${lang}_FLAGS_RELEASE_INIT " -O3 -DNDEBUG${_RTL_FLAGS}")
-    string(APPEND CMAKE_${lang}_FLAGS_RELWITHDEBINFO_INIT " -O2 -DNDEBUG${_DBG_FLAGS}${_RTL_FLAGS}")
+    string(APPEND CMAKE_${lang}_FLAGS_MINSIZEREL_INIT " -Os -DNDEBUG")
+    string(APPEND CMAKE_${lang}_FLAGS_RELEASE_INIT " -O3 -DNDEBUG")
+    string(APPEND CMAKE_${lang}_FLAGS_RELWITHDEBINFO_INIT " -O2 -DNDEBUG${_DBG_FLAGS}")
 
     # clang-cl accepts -RTC* flags but ignores them.  Simulate this
     # with the GNU-like drivers by simply passing no flags at all.
-- 
2.47.1

